File: apps/dma-insights/frontend/src/lib/recommendations.ts
/**
 * Recommendation detail typed hook.
 *
 * State-branch contract (driving the modal):
 *   - rec_id null              → query disabled
 *   - isLoading                → spinner row in the modal body
 *   - error/404                → "Couldn't load recommendation" empty
 *   - unresolved_count > 0     → amber "Pending review" banner above
 *                                the cited refs (cited ids that didn't
 *                                resolve render with strikethrough)
 *   - unresolved_count == 0    → happy: all citations clickable +
 *                                resolved name shown
 */
import { useQuery, type UseQueryResult } from "@tanstack/react-query";
import { apiGet } from "./api";

export interface CitedReference {
  kind: "feature" | "construct" | "agent";
  id: string;
  resolved: boolean;
  name: string | null;
}

export interface RecommendationDetail {
  id: string;
  rec_id: string;
  title: string;
  description: string;
  entity_display_id: string;
  target_subcap_ids: string[];
  platform_id: string | null;
  addressable_offerings: string[];
  uplift_per_pillar: Record<string, number> | null;
  effort_band: string | null;
  cited_features: CitedReference[];
  cited_constructs: CitedReference[];
  cited_agents: CitedReference[];
  unresolved_count: number;
  catalogue_version: string;
  // D4 DependencyMap: prerequisites (parsed at ingest) + the read-time
  // inverse unlocks. Both empty until the corpus is re-ingested.
  dependencies: { prerequisites: string[]; unlocks: string[] };
  /** Migration 048 (recommendations_detail.json / REC-NN.json ingest) —
   *  all additive; absent/empty until re-ingest fills them. `feature`
   *  is the concrete platform feature the rec ships; `phase` feeds the
   *  multi-phase roadmap. */
  feature?: string | null;
  phase?: number | null;
  /** E-IDs grounding the rec's root cause (modal "Root-cause evidence"
   *  tab). */
  root_cause_e_ids?: string[];
  /** Quantified expected outcomes {time, effort, metric, peer}. */
  outcomes?: {
    time?: string | null;
    effort?: string | null;
    metric?: string | null;
    peer?: string | null;
  } | null;
}

const UUID_RE =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

/**
 * Detail-fetch path builder (exported for the vitest ID contract).
 *
 * The stairstep curve and roadmap chevrons only hold the human-readable
 * `REC-NN` display code — codes are unique per run, not globally, so a
 * non-UUID id must carry the `display_id` scope for the backend to
 * resolve it (2026-07-06 drilldown-load fix; previously the code hit the
 * UUID-only lookup and 500'd, so those openers never loaded).
 */
export function recommendationDetailPath(
  recommendationId: string,
  displayId: string | null,
): string {
  const base = `/api/v1/recommendations/${encodeURIComponent(recommendationId)}`;
  if (!UUID_RE.test(recommendationId) && displayId) {
    return `${base}?display_id=${encodeURIComponent(displayId)}`;
  }
  return base;
}

/** Compose a RecommendationDetail from the client's baked roadmap row —
 * the pack-first fallback when the live API can't resolve the rec
 * (2026-07-06 deploy review: "I cannot load the recommendation from the
 * platform page" — pack rows carry rec_id forms, the endpoint knew only
 * uuids, and the hook had no snapshot tier at all). */
async function recFromSnapshot(
  recommendationId: string,
  displayId: string | null,
): Promise<RecommendationDetail | null> {
  if (!displayId) return null;
  const { pageSnapshot } = await import("./startup-pages");
  const rd = await pageSnapshot<{
    phases?: Array<{ recommendations?: Array<Record<string, unknown>> }>;
  }>(displayId, "platforms_roadmap");
  const want = recommendationId.toUpperCase();
  for (const ph of rd?.phases ?? []) {
    for (const r of ph.recommendations ?? []) {
      const rid = String(r.rec_id ?? r.id ?? "").toUpperCase();
      if (rid !== want && String(r.id ?? "") !== recommendationId) continue;
      const out = (r.outcomes ?? null) as RecommendationDetail["outcomes"];
      return {
        id: String(r.id ?? rid),
        rec_id: String(r.rec_id ?? rid),
        title: String(r.title ?? ""),
        description: String(r.description ?? r.body ?? ""),
        entity_display_id: displayId,
        target_subcap_ids: (r.target_subcap_ids as string[]) ?? [],
        platform_id: (r.platform_id as string) || null,
        addressable_offerings: [],
        uplift_per_pillar: null,
        effort_band: (r.effort_band as string) ?? null,
        cited_features: [], cited_constructs: [], cited_agents: [],
        unresolved_count: 0,
        catalogue_version: "",
        dependencies: { prerequisites: [], unlocks: [] },
        feature: (r.feature as string) ?? null,
        phase: (r.phase as number) ?? null,
        root_cause_e_ids: (r.root_cause_e_ids as string[]) ?? [],
        outcomes: out,
      };
    }
  }
  return null;
}

export function useRecommendationDetail(
  recommendationId: string | null,
  displayId: string | null = null,
): UseQueryResult<RecommendationDetail> {
  return useQuery({
    queryKey: ["recommendation", recommendationId, displayId],
    queryFn: async () => {
      try {
        // Live API first — REC-NN display codes carry the ?display_id=
        // scope (drilldown-load fix); UUIDs hit the bare detail path.
        return await apiGet<RecommendationDetail>(
          recommendationDetailPath(recommendationId as string, displayId),
        );
      } catch (err) {
        // Pack-first fallback (deploy review): compose from the client's
        // baked roadmap snapshot when the live API can't resolve the rec.
        const snap = await recFromSnapshot(String(recommendationId), displayId);
        if (snap) return snap;
        throw err;
      }
    },
    enabled: recommendationId !== null,
    staleTime: 60 * 1000,
  });
}

File: apps/dma-insights/backend/app/schemas/recommendations.py
"""Recommendation detail schema — drives D4's RecommendationModal."""
from __future__ import annotations

from pydantic import BaseModel, Field


class RecommendationNoteIn(BaseModel):
    """PUT body for the per-recommendation AE note. A blank/whitespace
    ``note`` DELETES the persisted row (see the PUT endpoint)."""

    note: str = Field(default="", max_length=20000)


class RecommendationNoteOut(BaseModel):
    """The durable, team-shared AE note for a (client, recommendation).

    ``note`` is "" when none is persisted (the GET endpoint returns an
    empty note object rather than 404 — the UI expects a shape, not a
    miss). ``author_email`` is the AE who last saved; ``updated_at`` is an
    ISO-8601 string (or null when there is no note)."""

    note: str = ""
    author_email: str | None = None
    updated_at: str | None = None


class CitedReference(BaseModel):
    """A single cited reference (feature / platform construct / agent).

    `resolved` tells the frontend whether the cited string ACTUALLY
    exists in the catalogue. Cited refs that don't resolve are surfaced
    so the AE knows the rec was generated with weaker grounding (and
    the analyst can decide whether to ship it).
    """
    kind: str        # "feature" | "construct" | "agent"
    id: str          # the cited string (e.g. "AF-LoanOriginator-01")
    resolved: bool
    name: str | None = None


class RecDependencies(BaseModel):
    """D4 DependencyMap: sibling recommendations that gate or are gated by
    this one. `prerequisites` are parsed at ingest from the source
    recommendation_validation.json; `unlocks` is the read-time inverse.
    Both empty (honest contextual-empty) until the corpus is re-ingested."""
    prerequisites: list[str] = Field(default_factory=list)
    unlocks: list[str] = Field(default_factory=list)


class RecommendationDetail(BaseModel):
    id: str
    rec_id: str
    title: str
    description: str
    entity_display_id: str
    target_subcap_ids: list[str] = Field(default_factory=list)
    platform_id: str | None = None
    addressable_offerings: list[str] = Field(default_factory=list)
    uplift_per_pillar: dict[str, float] | None = None
    effort_band: str | None = None
    cited_features: list[CitedReference] = Field(default_factory=list)
    cited_constructs: list[CitedReference] = Field(default_factory=list)
    cited_agents: list[CitedReference] = Field(default_factory=list)
    unresolved_count: int = 0
    catalogue_version: str
    dependencies: RecDependencies = Field(default_factory=RecDependencies)
    # Migration 048 (Part 7.2 — recommendations_detail.json / REC-NN.json
    # ingest). All additive — None/empty until re-ingest fills them.
    #
    # The concrete platform feature the rec ships (distinct from title).
    feature: str | None = None
    # Sequencing phase from the source rec files; feeds the multi-phase
    # roadmap (effort-band bucketing stays the fallback).
    phase: int | None = None
    # E-IDs grounding the rec's root cause — the modal's restored
    # "Root-cause evidence" tab.
    root_cause_e_ids: list[str] = Field(default_factory=list)
    # Quantified expected outcomes: {time, effort, metric, peer}.
    outcomes: dict | None = None
